lib/cretonne/src/verifier.rs

```diff
@@ -291,32 +301,42 @@ impl<'a> Verifier<'a> {
 
         // SSA form
         match dfg.value_def(v) {
-            // Value is defined by an instruction that exists and is inserted in an EBB.
             ValueDef::Res(def_inst, _) => {
+                // Value is defined by an instruction that exists.
                 if !dfg.insts.is_valid(def_inst) {
                     return err!(loc_inst,
                                 "{} is defined by invalid instruction {}",
                                 v,
                                 def_inst);
                 }
+                // Defining instruction is inserted in an EBB.
                 if self.func.layout.inst_ebb(def_inst) == None {
                     return err!(loc_inst,
                                 "{} is defined by {} which has no EBB",
                                 v,
                                 def_inst);
                 }
+                // Defining instruction dominates the instruction that uses the value.
+                if !self.domtree.dominates(def_inst, loc_inst, &self.func.layout) {
```

Comment:
IIRC, we defined dominance so an instruction dominates itself. It is not valid for an instruction to use its own result, though. Maybe that should be a separate check.